app/src/main/java/tech/jpco/qen/view/QenPage.kt

```diff
@@ -0,0 +1,154 @@
+        val (denormX, denormY) = denormalize(new.x, new.y)
+        val denormNew = new.copy(x = denormX, y = denormY)
+        if (lastPoint != null && denormNew.type != TouchEventType.TouchDown) {
+            lastPoint?.apply {
+                bufferCanvas.drawLine(x, y, denormNew.x, denormNew.y, paint)
+                if (shouldInvalidate) invalidate()
+            }
+        }
+        lastPoint = if (denormNew.type == TouchEventType.TouchUp) null else denormNew
+    }
+
+    fun clearPage() {
+        Log.d(TAG, "clearPage called")
+        bufferBitmap.eraseColor(Color.TRANSPARENT)
+        invalidate()
+    }
+
+    fun drawPage(list: List<DrawPoint>, newAR: Float) {
+        clearPage()
+        list.forEach { drawSegment(it, false) }
+        Log.d(TAG, "drawPage completing")
+        invalidate()
+    }
+
+    //comment-detritus from aborted aspect-ratio implementation
+    private fun normalize(x: Float, y: Float): Pair<Float, Float> =
+        /*if (ar < 1)*/ Pair(x / mWidth, y / mHeight)
+    /*else Pair(y / mHeight, x / mWidth)*/
+
+    private fun normDistance(pixels: Int) = pixels / (Math.sqrt((mHeight * mHeight + mWidth * mWidth).toDouble()))
+
+    //comment-detritus from aborted aspect-ratio implementation
+    private fun denormalize(relX: Float, relY: Float) =
+        /*if (ar < 1)*/ Pair(relX * mWidth, relY * mHeight)
+    /*else Pair(relY * mHeight, relX * mWidth)*/
+
+    private fun Observable<MotionEvent>.filterForActionAndTime(minMilli: Long): Observable<MotionEvent> =
+        publish { multiStream ->
+            Observable.merge(
+                multiStream
+                    .filter {
+                        it.action == MotionEvent.ACTION_DOWN ||
+                                it.action == MotionEvent.ACTION_UP
+                    },
+                multiStream
+                    .filter {
+                        it.action == MotionEvent.ACTION_MOVE
+                    }
+                    .throttleFirst(minMilli, TimeUnit.MILLISECONDS)
+
+            )
+        }
+
+
+    //scan() emits Pair(the last valid point, the new point IFF it's valid)
+    //what's better, .filter().map() or .flatmap() using Observable.empty() and Observable.just(it)?
+    private fun Observable<DrawPoint>.filterForMinDistance(tooClose: (Float, Float) -> Boolean): Observable<DrawPoint> =
+        scan(Pair<DrawPoint, DrawPoint?>(DrawPoint(Float.NaN, Float.NaN), null)) { standard, incoming ->
+            if (incoming.type == TouchEventType.TouchMove) {
+                val exX = standard.first.x
+                val exY = standard.first.y
+                if (exX != Float.NaN && exY != Float.NaN) {
+                    val xDif = incoming.x - exX
+                    val yDif = incoming.y - exY
+                    if (tooClose(xDif, yDif)) {
+                        return@scan Pair(standard.first, null)
+                    }
+                }
+
+            }
+            Pair(incoming, incoming)
```

Comment:
I'm actually particularly curious for your take on this particular workaround, and I guess I need to document better if it wasn't clearer why I have it the way I do!

Related to the comment on line 129, the purpose of applying the `scan` is to act as an extended variant of `distinctUntilChanged`. Instead of comparing the incoming point to the most recent one, it's comparing against the most recent point _that was valid_. The only way I could find at the time to do that kind of multi-event-spanning feedforward(?) loop without external state was to store it in a Pair, using standard.first for ongoing comparison and emitting a non-null standard.second only if it was a point I wanted to pass along to the VM and processor.

Now that I've played around with some other Rx constructions, maybe publish{} the receiver stream and use replay(1) to make the filtered stream available for comparison with new, not-yet-filtered events? Just spitballing...

Reply:
What you've done is definitely an augmented `distinctUntilChanged`, `scan` is the right operator to use. However, the chain can be simplified from `scan().filter().map()` to just a `scan().distinctUntilChanged()` if the `scan` operator just returns the new "previous" point. So if `tooClose` evaluates to `true`, return `standard`, and otherwise return `incoming`. Then you can slap on a `distinctUntilChanged()` since `DrawPoint` implements `==` and the behavior will be the same, but without the extra `Pair` creation.

Reply:
And this is the kind of suggestions that make me very glad I reached out to you! That is a much more elegant and self-evident solution to the problem.